File: python/merge_sort.py

def mergesort(coll,by=lambda x:x):
    if len(coll) > 1:
        hl = len(coll) // 2
        return merge_together(mergesort(coll[:hl],by),mergesort(coll[hl:],by),by)
    else:
        return coll


def merge_together(colla,collb,by):
    result = []
    ii = 0
    jj = 0
    while ii < len(colla) and jj < len(collb):
        if by(colla[ii]) >= by(collb[jj]):
            result.append(collb[jj])
            jj += 1
        else:
            result.append(colla[ii])
            ii += 1

    if ii == len(colla):
        result.extend(collb[jj:])
    else:
        result.extend(colla[ii:])
    return result

File: javascript/src/merge_sort.js
function mergesort(coll, by) {
    if (coll == undefined) {return [];}
    if (coll.length <= 1) {
        return coll;
    } else {
        if (arguments.length == 1) {
            by = function(x) {return x;};
        }
        var hl = Math.floor(coll.length / 2);
        return merge_together(mergesort(coll.slice(0,hl),by),
                              mergesort(coll.slice(hl),by),
                              by);
    }
}


function merge_together(colla,collb,by) {
    var result = [];
    var ii = 0;
    var jj = 0;
    while (ii < colla.length && jj < collb.length) {
        if (by(colla[ii]) >= by(collb[jj])) {
            result.push(collb[jj]);
            jj = jj + 1;
        } else {
            result.push(colla[ii]);
            ii = ii + 1;
        }
    }
    if (ii == colla.length) {
        result = result.concat(collb.slice(jj));
    } else {
        result = result.concat(colla.slice(ii));
    }
    return result;
}

exports = module.exports;
exports.mergesort = mergesort;


